linker.go

```diff
@@ -1,57 +1,149 @@
-func needSection(insns, section asm.Instructions) (bool, error) {
-	// A map of symbols to the libraries which contain them.
-	symbols, err := section.SymbolOffsets()
+// collectInstructions returns the instruction stream of all progs in order.
+func collectInstructions(progs []*ProgramSpec) asm.Instructions {
+	var out asm.Instructions
+
+	for _, prog := range progs {
+		out = append(out, prog.Instructions...)
+	}
+
+	return out
+}
+
+// collectFuncInfos returns the BTF func infos of all progs in order.
+func collectFuncInfos(progs []*ProgramSpec) ([]byte, error) {
+	if len(progs) == 0 {
+		return nil, nil
+	}
+
+	var off uint64
+
+	buf := bytes.NewBuffer(make([]byte, 0, binary.Size(&btf.FuncInfo{})*len(progs)))
+	for _, prog := range progs {
+		if err := prog.BTF.FuncInfo.Marshal(buf, off); err != nil {
+			return nil, fmt.Errorf("marshaling prog %s func info: %w", prog.Name, err)
+		}
+		off += prog.Instructions.Size()
+	}
+
+	return buf.Bytes(), nil
+}
+
+// collectLineInfos returns the BTF line infos of all progs in order.
+func collectLineInfos(progs []*ProgramSpec) ([]byte, error) {
+	if len(progs) == 0 {
+		return nil, nil
+	}
+
+	var off uint64
+
+	buf := bytes.NewBuffer(make([]byte, 0, binary.Size(&btf.LineInfo{})*len(progs)))
+	for _, prog := range progs {
+		if err := prog.BTF.LineInfos.Marshal(buf, off); err != nil {
+			return nil, fmt.Errorf("marshaling prog %s line infos: %w", prog.Name, err)
+		}
+		off += prog.Instructions.Size()
+	}
+
+	return buf.Bytes(), nil
+}
+
+// flatten returns a unique list of programs that need to be
+// included in spec's bytestream when inserting it into the kernel.
+//
+// It recursively collects spec's dependencies by stepping into its neighboring
+// ProgramSpecs. Each visited program is recorded to both avoid infinite
+// recursion and to prevent collecting the same program more than once.
+//
+// Always returns spec itself as the first entry. Returns nil when spec was
+// already visited.
+//
+// TODO: The stability of the order of the output matters,
+// this needs to be clear and tested.
+func (spec *ProgramSpec) flatten(visited map[*ProgramSpec]bool) []*ProgramSpec {
+	if visited == nil {
+		visited = make(map[*ProgramSpec]bool)
+	}
+
+	// This program and its dependencies were already collected.
+	if visited[spec] {
+		return nil
+	}
+
+	visited[spec] = true
+
+	// Always return the current program as the first one in the chain.
+	out := []*ProgramSpec{spec}
+
+	// Recurse into each neighbour and collect its neighbours,
+	// keeping track of visited nodes.
+	for _, dep := range spec.neighbours {
+		if dn := dep.flatten(visited); dn != nil {
+			out = append(out, dn...)
+		}
+	}
+
+	return out
+}
+
+// needProg checks insns for references to symbols in instruction stream dep.
+// Returns true if a reference is found from insns to dep.
+func needProg(insns, dep asm.Instructions) (bool, error) {
+	// A map of symbols to the instructions which contain them.
+	symbols, err := dep.SymbolOffsets()
```

Comment:
Aha, nice catch! This would be significant optimization, I'd been trying to think of ways to cut out some loops here and there.

Reply:
I think the `insns.Name()` helper I added in the latest commit could also be used here?

Reply:
I chose to match on the dependency's `insns.Name()` since `Reference` and `Symbol` are meant to correspond, not necessarily `ProgramSpec.Name`. If we implement linking in externally-provided program segments down the line, we might start calling `findReferences` from multiple places, and in that case what's in the instructions should prevail, not the program name.

I think maybe you or Arthur might disagree, but let's discuss this when we get there.